flink-filesystems/flink-gs-fs-hadoop/src/main/java/org/apache/flink/fs/gs/writer/GSRecoverableWriterCommitter.java

```diff
@@ -0,0 +1,189 @@
+        // only write the final blob if it doesn't already exist
+        Optional<GSBlobStorage.BlobMetadata> blobMetadata =
+                storage.getMetadata(recoverable.finalBlobIdentifier);
+        if (!blobMetadata.isPresent()) {
+            writeFinalBlob();
+        }
+
+        // clean up after successful commit
+        cleanupTemporaryBlobs();
+    }
+
+    @Override
+    public RecoverableWriter.CommitRecoverable getRecoverable() {
+        return recoverable;
+    }
+
+    /**
+     * Helper to compose an arbitrary number of blobs into a final blob, staying under the
+     * COMPOSE_MAX_BLOBS limit for any individual compose operation.
+     *
+     * @param sourceBlobIdentifiers The source blob ids to compose
+     * @param targetBlobIdentifier The target blob id for the composed result
+     */
+    private void composeBlobs(
+            List<GSBlobIdentifier> sourceBlobIdentifiers, GSBlobIdentifier targetBlobIdentifier) {
+        Preconditions.checkNotNull(sourceBlobIdentifiers);
+        Preconditions.checkArgument(sourceBlobIdentifiers.size() > 0);
+        Preconditions.checkNotNull(targetBlobIdentifier);
+
+        // split the source list into two parts; first, the ones we can compose in this operation
+        // (up to COMPOSE_MAX_BLOBS), and, second, whichever blobs are left over
+        final int composeToIndex =
+                Math.min(BlobUtils.COMPOSE_MAX_BLOBS, sourceBlobIdentifiers.size());
+        List<GSBlobIdentifier> composeBlobIds = sourceBlobIdentifiers.subList(0, composeToIndex);
+        List<GSBlobIdentifier> remainingBlobIds =
+                sourceBlobIdentifiers.subList(composeToIndex, sourceBlobIdentifiers.size());
+
+        // determine the resulting blob id for this compose operation. if this is the last compose,
+        // i.e. if there are no remaining blob ids, then the composed blob id is the originally
+        // specified target blob id. otherwise, we must create an intermediate blob id to hold the
+        // result of this compose operation
+        GSBlobIdentifier composedBlobId =
+                remainingBlobIds.isEmpty()
+                        ? targetBlobIdentifier
+                        : BlobUtils.generateTemporaryBlobIdentifier(
+                                recoverable.finalBlobIdentifier, options);
+
+        // compose the blobs
+        storage.compose(composeBlobIds, composedBlobId);
+
+        // if we have remaining blobs, add the composed blob id to the beginning of the list
+        // of remaining blob ids, and recurse
+        if (!remainingBlobIds.isEmpty()) {
+            remainingBlobIds.add(0, composedBlobId);
+            composeBlobs(remainingBlobIds, targetBlobIdentifier);
+        }
+    }
+
+    /**
+     * Writes the final blob by composing the temporary blobs and copying, if necessary.
+     *
+     * @throws IOException On underlying failure.
+     */
+    private void writeFinalBlob() throws IOException {
```

Comment:
nit:
```suggestion
    private void writeFinalBlob() {
```

Reply:
Done in [9e8472b](https://github.com/apache/flink/pull/15599/commits/9e8472bd88cbfa2cb7b606cf9c52f089aecabe6d).